### PR TITLE
Headless test runner

### PR DESCRIPTION
**Pull request by kateinoigakukun**
## Motivation 

`carton test --environment defaultBrowser` only works on macOS or X Window systems, so tests using browser APIs cannot be run on CI easily. Add a headless test runner based on WebDriver.

## Changed

- Add `WebDriverClient` module, which is a thin client of W3C WebDriver: https://w3c.github.io/webdriver/
- Add `--headless` option to `carton test`

### CODE REVIEW

**Comment by MaxDesiatov** [comment]
Is this new option compatible with other `--environment` values? If not, maybe it should be implemented as `--environment webdriver`?

**Comment by kateinoigakukun** [comment]
@MaxDesiatov I made it as a new flag instead of a `--environment` variant because I thought `--environment` is designed to accept specific browser names in the future like `--environment firefox` and then `--headless` should be able to be used with `defaultBrowser` and other browser values.

**Comment by MaxDesiatov** [comment]
That makes sense, thanks!

File: README.md [hunk 1]
```diff
@@ -93,6 +93,9 @@ either: `wasmer`, `node` or `defaultBrowser`. Code that depends on
 [JavaScriptKit](https://github.com/swiftwasm/JavaScriptKit) should pass either `--environment node` or
 `--environment defaultBrowser` options, depending on whether it needs Web APIs to work. Otherwise
 the test run will not succeed, since JavaScript environment is not available with `--environment wasmer`.
+If you want to run your test suite on CI or without GUI but on browser, you can pass `--headless` flag.
+It enables [WebDriver](https://w3c.github.io/webdriver/)-based headless browser testing. Note that you
+need to install a WebDriver in `PATH` before running tests.
```
Comment (MaxDesiatov) [inline README.md:98]:
```suggestion
need to install a WebDriver executable in `PATH` before running tests.
```

File: Sources/CartonCLI/Commands/Test.swift [hunk 2]
```diff
@@ -41,6 +45,12 @@ struct Test: AsyncParsableCommand {
   )
   private var environment = Environment.wasmer
 
+  // It implemented as a separate flag instead of a `--environment` variant because `--environment`
+  // is designed to accept specific browser names in the future like `--environment firefox`. 
+  // Then `--headless` should be able to be used with `defaultBrowser` and other browser values.
```
Comment (MaxDesiatov) [inline Sources/CartonCLI/Commands/Test.swift:50]: Sorry, I wanted to suggest a doc comment, but used a regular comment delimiter for the previous suggestion.
```suggestion
  /// It is implemented as a separate flag instead of a `--environment` variant because `--environment`
  /// is designed to accept specific browser names in the future like `--environment firefox`. 
  /// Then `--headless` should be able to be used with `defaultBrowser` and other browser values.
```

File: Sources/CartonCLI/Commands/TestRunners/BrowserTestRunner.swift [hunk 3]
```diff
@@ -12,39 +12,165 @@
...
 import TSCBasic
+import WebDriverClient
 
 private enum Constants {
   static let entrypoint = Entrypoint(fileName: "test.js", sha256: testEntrypointSHA256)
 }
 
+enum BrowserTestRunnerError: Error, CustomStringConvertible {
+  case invalidRemoteURL(String)
+  case failedToFindDriver
```
Comment (MaxDesiatov) [inline Sources/CartonCLI/Commands/TestRunners/BrowserTestRunner.swift:31]: nit: let's make naming more explicit here
```suggestion
  case failedToFindWebDriver
```

File: Sources/CartonCLI/Commands/TestRunners/BrowserTestRunner.swift [hunk 4]
```diff
@@ -12,39 +12,165 @@
...
+
+  var description: String {
+    switch self {
+    case let .invalidRemoteURL(url): return "Invalid remote URL: \(url)"
+    case .failedToFindDriver:
+      return """
+      Failed to find WebDriver executable or remote URL to a running driver process.
+      Please make sure that you satisfied one of the followings (smaller item number lower priority):
```
Comment (MaxDesiatov) [inline Sources/CartonCLI/Commands/TestRunners/BrowserTestRunner.swift:39]: nit: I'd rephrase it this way:
```suggestion
      Please make sure that you are satisfied with one of the followings (in order fo priority):
```

File: Sources/CartonCLI/Commands/TestRunners/BrowserTestRunner.swift [hunk 5]
```diff
@@ -12,39 +12,165 @@
...
   let testFilePath: AbsolutePath
   let host: String
   let port: Int
+  let headless: Bool
   let manifest: Manifest
   let terminal: InteractiveWriter
+  let eventLoopGroup = MultiThreadedEventLoopGroup(numberOfThreads: 1)
+  let httpClient: HTTPClient
+
+  internal init(
```
Comment (MaxDesiatov) [inline Sources/CartonCLI/Commands/TestRunners/BrowserTestRunner.swift:58]: nit: the convention is not to use `internal` as it's the implicit default
```suggestion
  init(
```

File: Sources/WebDriverClient/WebDriverClient.swift [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+    return url.absoluteString
+  }
+
+  private static func makeRequestBody<R: Encodable>(_ body: R) throws -> HTTPClient.Body {
+    let encoder = JSONEncoder()
+    return try HTTPClient.Body.data(encoder.encode(body))
+  }
+
+  public func goto(url: String) async throws {
+    struct Response: Decodable {}
```
Comment (MaxDesiatov) [inline Sources/WebDriverClient/WebDriverClient.swift:95]: this inner `Response` struct doesn't seem to be used anywhere?

Comment (kateinoigakukun) [reply Sources/WebDriverClient/WebDriverClient.swift]: Good catch

**Review by MaxDesiatov — APPROVED** [review]
Great work, thanks! 👏
